### ISSUE TITLE
Make authentication more flexible

### ISSUE THREAD

**Issue by chrissnell** [jamiealquiza/polymur#38]
I was super excited to find polymur tonight.  Exactly what I needed: a way to efficiently move Graphite metrics from a remote site over HTTPS.  However, the authentication system has the unfortunate dependency on Consul.  We're a CoreOS/Kubernetes shop and don't run Consul anywhere--we run etcd.  So, this is probably too big of a burden to run polymur as-is.   

I'm trying to understand why you chose to validate keys in a K/V store versus using the very-standard method of having clients present signed certs and having the server-side validate these certs.  This would eliminate the dependency upon a service that not everybody runs.

Another option I'm considering is adapting polymur to use etcd instead of Consul but this still feels a bit funky to me.

Thoughts?

**Comment by chrissnell** [comment]
Did some more digging.  This is what I have in mind:

   http://www.levigross.com/2015/11/21/mutual-tls-authentication-in-go/

I also looked into swapping out etcd for consul or just making your key management into an Go interface a doing an etcd implementation but I feel like that's the wrong solution--custom authentication schemes almost always are.  TLS auth will do what is needed and do it in a very standard way.  As for polymur's optional key prefixing feature, that could be modified to use the presented certificate's CN as the optional prefix.

**Comment by jamiealquiza** [comment]
There's a few reasons it's setup this way:

- Baking auth is far less daunting than that phrase of "don't roll your own crypto"; the security/crypto is still just riding on HTTPS.
- The API keys model makes it easy to create an arbitrary number of keys without touching the gateways and quickly perform revocation on a per-key basis (e.g. not everyone will use Polymur-Gateway for their own infra; we supported arbitrary numbers of remote sites that we did not manage, and therefore may not trust).
- It was desirable to separate authorization and transport security. Polymur-Gateway isn't necessarily terminating the crypto (see the diagrams where gateways may be behind an SSL terminating ELB). The current API key scheme moves auth into the request parameters rather than being bound to the transport layer in anyway. But yeah, this is achievable in more ways than what Polymur is doing.
- The key model removes physical artifacts (certs/keys) from the boxes and replaces them with runtime config. While handling artifacts can simply be polished off in your config management, I've moved on from that systems management philosophy. The goal was to either remove or push state off the app and into the environment. It makes arbitrarily killing/provisioning Polymur-Gateway boxes easier (in the context of that production env).
- API Keys also remove an artifact on the client side. The only config a client needs is the gateway address and the key as runtime config, which can be fetched from Consul, Etcd, etc. (though not natively, yet).
- Consul in particular was used because it was the existing environment config facility in the large production environment Polymur was designed in. I would be interested in modularizing the API key backend to support more services - but then again, Consul is quite easy to setup. Even a single node setup could suffice since running gateways will continue to operate with a consul loss (key revocation / creation just won't propagate the gateways until Consul returns).

That said, I’m all about improving user experience, so I putting aside all considerations above, supporting more than just Consul would be a great idea. I also like the idea of possibly adding configuration flexibility to the namespace prefix.

**Comment by chrissnell** [comment]
OK, fair enough.  For us, Kubernetes makes it easy enough to package
keys+certs in the environment (via kube's Secret object) but that might not
be true for other users.

I started work on implementing cert checking but I'm going to put that on
hold and try breaking out API key checking into an interface and make an
etcd implementation.  That's "good enough" for us.

**Comment by jamiealquiza** [comment]
Great, would be interesting to see. I still openly support the idea of making key sync more generic in both:

- what backends does it support?
- how does it function?

**Comment by chrissnell** [comment]
So I'm immediately running into a little roadblock: I need to get configuration options down to the etcd connector; these options include host, port, cert, key, directory prefix for key storage, etc.  Normally, in my software, I ingest a YAML config file and populate a Config object that gets passed around by reference to everything that needs to access config data.  It cleans up the code a little bit because instead of passing things like options.etcdHostname to the keysync goroutine, you just pass *Config.  It's especially useful when that keysync goroutine becomes an implementation of an interface and has a bunch of different config variables for each different implementation (consulHostname, etcdHostname, mysqlHostname, ...).  

So, I'm stuck.  I don't have the time to retrofit all of your code and introduce a Config and a YAML-based config file.  I don't even know if that's what you want.  But, if I start passing a bunch of etcd-specific options to the keysync, it's going to get messy, fast.

Thoughts?

**Comment by chrissnell** [comment]
Here's an example of how I do it.

The config.go parser:
   https://github.com/chrissnell/crabby/blob/master/config.go

An example config.yaml:
   https://github.com/chrissnell/crabby/blob/master/config.yaml

**Comment by jamiealquiza** [comment]
Yeah, the keysync code is still something like a first-iteration. At the least, I usually would pass around a generic option type (or in your reference, *Config), e.g. [here](https://github.com/jamiealquiza/polymur/blob/master/cmd/polymur-gateway/main.go#L41). I'll give the examples a read tomorrow and re-look at the keysync code and see what best way to make it more generic and extensible seems to make sense.

Also, it looks like you're wanting to push all config into etcd, beyond API key sync, right?

**Comment by chrissnell** [comment]
Nah, I don't keep any configuration in etcd.  I use environment variables and/or command-line args for simple stuff, and YAML for more complex configurations.  

If I was using YAML for polymur, the config might have a section like this:

```yaml

authentication:
    enabled: true
    keystore:
        etcd:
            hostname: myetcd-server.domain
            port: 2379
            cert: etcd-client.pem
            key: etcd-client.key
            key-prefix: /techops/polymur
```

...and a similar section under `keystore` for Consul, etc.
